### PR TITLE
Sheffield | 26-Java-June | Hassan Osman | Sprint 3 | Prep/Backlog: Automated Testing

### PR DESCRIPTION
**Pull request by HassanOHOsman**
Self checklist

- [X] I have titled my PR with Region | Cohort | FirstName LastName | Sprint | Assignment Title
- [X] My changes meet the requirements of the task
- [X] I have tested my changes
- [X] My changes follow the [style guide](https://curriculum.codeyourfuture.io/guides/reviewing/style-guide/)

## Changelist

This PR contains answers to the prep exercises for the automated testing

## Questions

1. Regarding the  failure messages, do they have to be written specifically for the test case? or generically?

### CODE REVIEW

File: .idea/Module-Java.iml
Comment (jacknewman1995) [inline .idea/Module-Java.iml:1]: There are a bunch of non Java files here that i don't think should have been pushed

Comment (jacknewman1995) [reply .idea/Module-Java.iml:1]: Looks like they are intellij generated files. Definitely don't want to push them - in real life including them in a PR would mean that if merged the next person to pull the changes would be stuck with your IJ files

File: exercise 1.1 - 1.2/src/com/example/Calculator.java [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+    public int subtract(int a, int b) {
+        return a - b;
+    }
+
+    public int multiply(int a, int b) {
+        return a * b;
+    }
+
+    public int divide(int a, int b) {
+        if (b == 0) {
```
Comment (jacknewman1995) [inline exercise 1.1 - 1.2/src/com/example/Calculator.java:17]: Nice

File: exercise 1.1 - 1.2/src/com/example/CalculatorTest.java [hunk 2]
```diff
@@ -0,0 +1,50 @@
+package com.example;
+
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.DisplayName;
+import org.junit.jupiter.api.Test;
+import static org.junit.jupiter.api.Assertions.*;
+
+class CalculatorTest {
+
+    Calculator calculator;
```
Comment (jacknewman1995) [inline exercise 1.1 - 1.2/src/com/example/CalculatorTest.java]: Does this need to be package visible?

File: exercise 1.1 - 1.2/src/com/example/CalculatorTest.java [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+
+    @BeforeEach
+    void setUp() {
```
Comment (jacknewman1995) [inline exercise 1.1 - 1.2/src/com/example/CalculatorTest.java]: You don't need a beforeeach here, you can just create it as a field and you'll get a new one each time. Junit creates a new instance of the class per test method

File: exercise 1.1 - 1.2/src/com/example/CalculatorTest.java [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+        calculator =  new Calculator();
+    }
+
+    @Test
+    void testAdd() {
+        assertEquals(10, calculator.add(5, 5), "Addition of 2 values should equal the expected value");
```
Comment (jacknewman1995) [inline exercise 1.1 - 1.2/src/com/example/CalculatorTest.java:14]: Note in earlier versions of junit the message is the first parameter - this ordering is new to me!

File: exercise 1.1 - 1.2/src/com/example/CalculatorTest.java [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+    }
+
+    @Test
+    void testDivide() {
+        assertEquals(20, calculator.divide(100, 5), "dividing value1 by value2 should equal the expected value");
```
Comment (jacknewman1995) [inline exercise 1.1 - 1.2/src/com/example/CalculatorTest.java:19]: Don't feel like you always have to write a message - for tests with only one assertion it's pretty obvious what it's testing, but adding a message can be useful when you've got a bunch of similar assertions e.g. testing properties of an object, and want a clearer error than just "assertion error"

File: exercise 1.1 - 1.2/src/com/example/CalculatorTest.java [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+    }
+
+    @Test
+    void testDivideByZeroThrows() {
+        assertThrows(IllegalArgumentException.class, ()-> calculator.divide(50, 0), "dividing by 0 should throw an exception");
+    }
+
+    @Test
+    @DisplayName("Method returns the square of an integer")
```
Comment (jacknewman1995) [inline exercise 1.1 - 1.2/src/com/example/CalculatorTest.java:28]: What's display name that's new to me?

File: exercise 1.1 - 1.2/src/com/example/CalculatorTest.java [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+    void testSquare() {
+        assertEquals(64, calculator.square(7), "when a value is squared it should equal to the expected result");
```
Comment (jacknewman1995) [inline exercise 1.1 - 1.2/src/com/example/CalculatorTest.java:30]: Does this test pass..?

File: exercise 1.1 - 1.2/src/com/example/CalculatorTest.java [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+    }
+
+    @Test
+    @DisplayName("Method returns the square root of an integer")
+    void testSquareRoot() {
+        assertEquals(5, calculator.squareRoot(25), "The square root of a value should equal to the expected result");
```
Comment (jacknewman1995) [inline exercise 1.1 - 1.2/src/com/example/CalculatorTest.java]: Both error messages being the same isn't very helpful as it won't tell the user which assertion has failed. Either make them specific or just use the default message

File: exercise 1.1 - 1.2/src/com/example/Main.java [hunk 9]
```diff
@@ -0,0 +1,13 @@
+//TIP To <b>Run</b> code, press <shortcut actionId="Run"/> or
```
Comment (jacknewman1995) [inline exercise 1.1 - 1.2/src/com/example/Main.java]: Shouldnt have been pushed

File: exercise 1.3/src/StringUtilities.java [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+
+        String reversedString = new StringBuilder(s).reverse().toString();
+
+        return s.equals(reversedString);
+    }
+
+
+    public List<String> splitCommaSeparatedLine(String s) {
+
+        String[] stringsArray = s.split("\\s*,\\s*");
```
Comment (jacknewman1995) [inline exercise 1.3/src/StringUtilities.java:15]: Does this work as expected?

File: exercise 1.3/src/StringUtilitiesTest.java [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+
+import org.junit.jupiter.params.provider.CsvSource;
+import org.junit.jupiter.params.provider.MethodSource;
+import org.junit.jupiter.params.provider.Arguments;
+
+
+
+public class StringUtilitiesTest {
+
+    StringUtilities stringUtilities;
```
Comment (jacknewman1995) [inline exercise 1.3/src/StringUtilitiesTest.java]: Private

File: exercise 1.3/src/StringUtilitiesTest.java [hunk 12]
```diff
@@ -0,0 +1,66 @@
...
+    void testIsPalindrome1(){
+        assertTrue(stringUtilities.isPalindrome("mom"), "this word should be palindrome");
+        assertTrue(stringUtilities.isPalindrome("B"), "this word should be palindrome"); //edge case
+        assertTrue(stringUtilities.isPalindrome(""), "this word should be palindrome");  // edge case
+    }
+
+    @ParameterizedTest
+    @CsvSource({"RADAR", "dead", "a"})
+    @DisplayName("Test if various input strings were palindrome all at once using @ParameterizedTest and @CsvSource")
+    void testIsPalindrome2(String s) {
```
Comment (jacknewman1995) [inline exercise 1.3/src/StringUtilitiesTest.java:32]: Some junit 5 annotations here that I'm not familiar with - are the parameters to use Radar, dead and a? If so, why does this test pass?

File: exercise 1.3/src/StringUtilitiesTest.java [hunk 13]
```diff
@@ -0,0 +1,66 @@
...
+    }
+
+    @ParameterizedTest
+    @MethodSource("isPalindromeDate")
+    @DisplayName("")
+    void testIsPalindrome3(String s) {
+        assertTrue(stringUtilities.isPalindrome(s));
+    }
+
+    static Stream<Arguments> isPalindromeDate() {
```
Comment (jacknewman1995) [inline exercise 1.3/src/StringUtilitiesTest.java]: Have you seen streams before?

File: exercise 1.3/src/StringUtilitiesTest.java [hunk 14]
```diff
@@ -0,0 +1,66 @@
...
+                Arguments.of("hahaha")
+        );
+    }
+
+
+    @Test
+    void splitCommaSeparatedLine(){
+        assertEquals(List.of("Jan", "Feb", "Mar", "Apr"), stringUtilities.splitCommaSeparatedLine("Jan, Feb, Mar, Apr"), "Comma-separated string should return a list of those strings");
+        assertEquals(List.of("Spring - Summer - Autumn - Winter"), stringUtilities.splitCommaSeparatedLine("Spring - Summer - Autumn - Winter"), "Comma-separated string should return a list of those strings"); //edge case
+        assertEquals(List.of("  ", "  ", "  ", "  "), stringUtilities.splitCommaSeparatedLine("  ,  ,  ,  "), "Comma-separated string should return a list of those strings");// edge case
```
Comment (jacknewman1995) [inline exercise 1.3/src/StringUtilitiesTest.java]: How come the space is preserved here? Looks like your split method should trim it?

File: exercise1.4/src/main/java/org/example/Calculator.java [hunk 15]
```diff
@@ -0,0 +1,34 @@
+package org.example;
+
```
Comment (jacknewman1995) [inline exercise1.4/src/main/java/org/example/Calculator.java:2]: Duplicate class?

File: exercise1.4/src/main/java/org/example/StringUtilities.java [hunk 16]
```diff
@@ -0,0 +1,31 @@
...
+
+        return stringsList;
+
+    }
+}
+
+/*
+Question and Answer:
+Q1. Note that source and test classes should be in separate folders (src/main/java, src/test/java). What do you notice if they are all in src/main/java?
+A1. This creates confusion, specially when it comes to the responsiblity/role of each class.
```
Comment (jacknewman1995) [inline exercise1.4/src/main/java/org/example/StringUtilities.java:30]: Often only the source folders are shipped to clients, so we don't want to include any test classes. Also, when you run all tests in a project it will only run tests under /test, so any tests in main wont run and you'll lose coverage!

**Comment by HassanOHOsman** [comment]
> In general the code looks great, nice one! Looks like there's lots of files included here that shouldn't be, e.g. example files and IJ generated files. Did you git add . or something? Can you remove the files that don't need review please?

I will do this now. Apologies. Whenever I commit, I always get a warning, so perhaps something in the setting that doesn't communicate to me those other files are being committed too. I don't know, I feel it's lacking here compared to VSCode for example. 😅

Anyway, I will look into it and get them removed. Thanks, Jack 😀

File: exercise 1.1 - 1.2/src/com/example/CalculatorTest.java (hunk 2 above)
Comment (HassanOHOsman) [reply exercise 1.1 - 1.2/src/com/example/CalculatorTest.java]: I don't think this class needs to be publicly visible specially that no other classes in other packages need to access it. So package-private is ideal here.

File: exercise 1.1 - 1.2/src/com/example/CalculatorTest.java (hunk 3 above)
Comment (HassanOHOsman) [reply exercise 1.1 - 1.2/src/com/example/CalculatorTest.java]: This is great to know!!!! I wonder why the examples in the articles for the automated testing do it this way, then? My first thought was that perhaps it was a new Junit5 feature, and that older Junit versions would do it this way, to get a fresh instance prior to each test. However, as per my research, @BeforeEach doesn't exist in versions before Junit5.

Without digressing here, could I know generic examples as to when @BeforeEach would be useful, please? Thank you :)

File: exercise 1.1 - 1.2/src/com/example/CalculatorTest.java (hunk 4 above)
Comment (HassanOHOsman) [reply exercise 1.1 - 1.2/src/com/example/CalculatorTest.java:14]: Interesting! Thanks for the heads-up :)

File: exercise 1.3/src/StringUtilitiesTest.java (hunk 12 above)
Comment (HassanOHOsman) [reply exercise 1.3/src/StringUtilitiesTest.java:32]: yes, the parameters for this test were "Radar," "dead" and "a". "dead" fails, while "Radar" and "a" pass. This is because the order is not a palindrome, while the latter were.

I've attached a snippet of the results for this test for your reference.
Screenshot 2026-08-06 at 06 57 22

File: exercise 1.3/src/StringUtilitiesTest.java (hunk 13 above)
Comment (HassanOHOsman) [reply exercise 1.3/src/StringUtilitiesTest.java]: No, that was my first encounter. The exercise was similar to an example exercise in one of the reading materials.

File: exercise 1.3/src/StringUtilitiesTest.java (hunk 14 above)
Comment (HassanOHOsman) [reply exercise 1.3/src/StringUtilitiesTest.java]: It's the highlighted regex in the attachment below that does this trick.
Screenshot 2026-08-06 at 07 10 28

File: exercise1.4/src/main/java/org/example/Calculator.java (hunk 15 above)
Comment (HassanOHOsman) [reply exercise1.4/src/main/java/org/example/Calculator.java:2]: I copies this class from the previous exercise, as one of the goals of exercise 4 was to maven-structure the previous work.

File: .idea/Module-Java.iml
Comment (HassanOHOsman) [reply .idea/Module-Java.iml:1]: I understand how hard it can be. I apologise for repeating the same mistake. I've just ignored the files and folders in questions. Thank you very much, Jack :)

File: exercise 1.1 - 1.2/src/com/example/CalculatorTest.java (hunk 5 above)
Comment (HassanOHOsman) [reply exercise 1.1 - 1.2/src/com/example/CalculatorTest.java:19]: Very useful! Thank you :)

File: exercise 1.1 - 1.2/src/com/example/CalculatorTest.java (hunk 6 above)
Comment (HassanOHOsman) [reply exercise 1.1 - 1.2/src/com/example/CalculatorTest.java:28]: @DisplayName adds more clarity as it what a method does. Think of it like making a method's identifier more human readable. 

In this example, the test method identifier is testSquare, it's not the most meaningful name  so having @DisplayName can communicate things better, specially that spaces are allowed. 

On a side note, I watched a YouTube video and the person said if your colleague named a test method horribly and you want be passing-aggressive about it, use @DisplayName. I found it funny lol 🤣

File: exercise 1.1 - 1.2/src/com/example/CalculatorTest.java (hunk 7 above)
Comment (HassanOHOsman) [reply exercise 1.1 - 1.2/src/com/example/CalculatorTest.java:30]: It fails. I've attached a screenshot for your review.
Screenshot 2026-08-06 at 08 17 37

File: exercise 1.1 - 1.2/src/com/example/CalculatorTest.java (hunk 8 above)
Comment (HassanOHOsman) [reply exercise 1.1 - 1.2/src/com/example/CalculatorTest.java]: This actually answers the question I left along with my PR check list and description. I've just updated this and used scenario-specific error messages, instead. Thank you! :)

File: exercise 1.1 - 1.2/src/com/example/Main.java (hunk 9 above)
Comment (HassanOHOsman) [reply exercise 1.1 - 1.2/src/com/example/Main.java]: I understand. When i chose to have my project Maven-structured on intelliJ, the Main.class file with its main method were auto-generated. I was oblivious to that. However, I just got rid of it, altogether, since I don't need it for exercises 1.1-1.2 anyway.

File: exercise 1.3/src/StringUtilities.java (hunk 10 above)
Comment (HassanOHOsman) [reply exercise 1.3/src/StringUtilities.java:15]: Yes, it does. However, doesn't work for the following edge case I came up with: `assertEquals(List.of("  ", "  ", "  ", "  "), stringUtilities.splitCommaSeparatedLine("  ,  ,  ,  ")` 

I've attached a screenshot as evidence that other cases passed.
Screenshot 2026-08-06 at 08 46 38

File: exercise 1.3/src/StringUtilitiesTest.java (hunk 11 above)
Comment (HassanOHOsman) [reply exercise 1.3/src/StringUtilitiesTest.java]: Will do, thank you! Out of interest, can it also be package-private?

File: exercise1.4/src/main/java/org/example/StringUtilities.java (hunk 16 above)
Comment (HassanOHOsman) [reply exercise1.4/src/main/java/org/example/StringUtilities.java:30]: That's actually brilliant!!!!!! Thank you very much for the great info :)

What are the cases where you'd include tests in main? I couldn't imagine clients requesting this. Why would they care if that was the case?

File: exercise 1.1 - 1.2/src/com/example/CalculatorTest.java (hunk 7 above)
Comment (jabeywick) [reply exercise 1.1 - 1.2/src/com/example/CalculatorTest.java]: Should the test be failing?

File: exercise 1.1 - 1.2/src/com/example/CalculatorTest.java (hunk 3 above)
Comment (jabeywick) [reply exercise 1.1 - 1.2/src/com/example/CalculatorTest.java]: I personally haven't seen them a huge amount with unit tests. It probably makes more sense to use when you're doing something more complicated at the start of each test as opposed to creating a new object.
